Add new scheduler of load in fe

EmmyMiao87 opened the pull request:
1. New scheduler only support the broker load now.
2. The stage of load consist of PENDING -> LOADING -> FINISHED
3. The LoadScheduler will divide job into a pending task. There are preparations that need to be done on pending task.
4. OnPendingTaskFinished will be invoked after pending task. It is used to submit the loading task which is created based on attachment of pending task.
5. OnLoadingTaskFinished will be invoked after loding task. It is used to record the commit info and commit txn when all of task has been finished.
.

EmmyMiao87 committed:
Add new scheduler of load in fe

1. New scheduler only support the broker load now.
2. The stage of load consist of PENDING -> LOADING -> FINISHED
3. The LoadScheduler will divide job into a pending task. There are preparations that need to be done on pending task.
4. OnPendingTaskFinished will be invoked after pending task. It is used to submit the loading task which is created based on attachment of pending task.
5. OnLoadingTaskFinished will be invoked after loding task. It is used to record the commit info and commit txn when all of task has been finished.
.

morningman commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadManager.java:
```diff
@@ -0,0 +1,104 @@
+            new MasterTaskExecutor(Config.load_pending_thread_num_high_priority);
+    private MasterTaskExecutor loadingTaskExecutor = new MasterTaskExecutor(10);
+
+    private ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
```
What is this lock for?

EmmyMiao87 replied:
To protect the atomicity of (idToLoadJob, dbIdToLabelToLoadJobs)

morningman commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadingTaskAttachment.java:
```diff
@@ -0,0 +1,58 @@
+
+public class BrokerLoadingTaskAttachment implements LoadLoadingTaskAttachment {
+
+    private Map<String, Long> fileMap;
```
Plz add comment to explain the key and value of the map

EmmyMiao87 replied:
I will remove it later.

morningman commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadScheduler.java:
```diff
@@ -0,0 +1,90 @@
+    private LoadManager loadManager;
+
+    public LoadScheduler(LoadManager loadManager) {
+        super();
```
Better adding a thread name for debugging easily

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadTaskCallback.java:
```diff
@@ -0,0 +1,33 @@
+
+package org.apache.doris.load.loadv2;
+
+public interface LoadTaskCallback {
```
wen can make this callback more general.

I think that define only `onTaskFinished` and `onTaskFailed` is enough and give job more options

EmmyMiao87 commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadLoadingTask.java:
```diff
@@ -0,0 +1,167 @@
+        this.callback = callback;
+    }
+
+    public void init(List<List<TBrokerFileStatus>> fileStatusList, int fileNum) throws UserException {
```
The db lock has been held in the BrokerLoadJob.

EmmyMiao87 committed:
Combine pendingTask and loadingTask into loadTask

1. The load task callback include two methods: onTaskFinished, onTaskFailed

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,229 @@
+                          attachment.getFileNumByTable(tableId));
+                // Add tasks into list and pool
+                tasks.add(task);
+                Catalog.getCurrentCatalog().getLoadManager().submitTask(task);
```
Can you pass loadManager when this object is created?
we should avoid using getInstance() in our code

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,229 @@
+            onPendingTaskFinished((BrokerPendingTaskAttachment) attachment);
+        } else if (attachment instanceof BrokerLoadingTaskAttachment) {
+            onLoadingTaskFinished((BrokerLoadingTaskAttachment) attachment);
+        }
```
Do you have some `state` or something to indicate whether or not this call is valid in your state machine?

For example this function can be called concurrently? or called multiple times for same task?

EmmyMiao87 replied:
Situation 1 (BrokerPendingTask ):There is only one pending task in BrokerLoad. So this function will not be called concurrently when pending task.
Situation 2 (LoadingTask): I add the writelock before onLoadingTaskFinished. So the loading task will be serial executed.

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,229 @@
+
+            // begin commit txn when all of loading tasks have been finished
+            if (tasks.size() == tasks.stream()
+                    .filter(entity -> entity.isFinished()).count()) {
```
why not use a set to record which task has finished?

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,229 @@
+
+        Database db = null;
+        try {
+            getDb();
```
db = getDb()?

EmmyMiao87 committed:
Fix defect of task callback

EmmyMiao87 committed:
Fix some error

EmmyMiao87 committed:
Add a separate LoadTaskScheduler

EmmyMiao87 committed:
Add txn callback of load job

1. isCommittting is true when beforeCommitted in txn
2. job could not be cancelled when isCommitting is true
3. job will be finished after txn is visible
4. old job will be cleaned when (CurrentTS - FinishedTs) / 1000 > Config.label_keep_seconds
5. LoadTimeoutChecker is performed to cancel timeout job

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,265 @@
+                                     .build());
+                    cancelJobWithoutCheck(FailMsg.CancelType.ETL_RUN_FAIL,
+                                          "Unknown table(" + tableId + ") in database(" + db.getFullName() + ")");
+                    return;
```
throw Exception?

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadJob.java:
```diff
@@ -0,0 +1,265 @@
+        loadStartTimestamp = System.currentTimeMillis();
+    }
+
+    private void allocateLoadingTask(Database db, BrokerPendingTaskAttachment attachment) throws UserException {
```
```suggestion
    private void createLoadingTask(Database db, BrokerPendingTaskAttachment attachment) throws UserException {
```

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/LoadJobScheduler.java:
```diff
@@ -0,0 +1,92 @@
+
+            // schedule job
+            try {
+                loadJob.scheduleJob();
```
scheduler call `scheduleJob`? a little strange

EmmyMiao87 committed:
Poll job from needScheduleJobsQueue

EmmyMiao87 committed:
Change to begin job in scheduler

imay commented on fe/src/main/java/org/apache/doris/load/loadv2/BrokerLoadPendingTask.java:
```diff
@@ -0,0 +1,85 @@
+                }
+                fileStatusList.add(fileStatuses);
+                for (TBrokerFileStatus fstatus : fileStatuses) {
+                    if (LOG.isDebugEnabled()) {
```
put this judge out of `for` block?

EmmyMiao87 committed:
Fix some mistake
